content/blog/copilot-lessons/index.md

```diff
@@ -0,0 +1,127 @@
+---
+title: "Engineering Lessons from Building Pulumi Copilot"
+date: 2024-12-12T13:56:13-05:00
+draft: false
+meta_desc: Learn key engineering lessons from building Pulumi Copilot, including how to minimize LLM workload, validate outputs, and deal with hallucination.
+meta_image: meta.png
+authors:
+    - artur-laksberg
+    - adam-gordon-bell
+tags:
+    - Copilot
+    - ai
+    - iac
+social:
+    twitter: 🚀 Building Pulumi Copilot taught us key AI engineering insights – Minimize LLM workload, validate outputs rigorously & use "skills" for modular tasks. Plus, sometimes when AI hallucinates features, it reveals exactly what users want! 
+    linkedin: 🚀 Building Pulumi Copilot taught us key AI engineering insights – Minimize LLM workload, validate outputs rigorously & use "skills" for modular tasks. Plus, sometimes when AI hallucinates features (like a "--force" flag), it reveals exactly what users want! 
+---
+**Artur Laksberg is on the Copilot team here at Pulumi. Copilot is an AI assistant for cloud infrastructure. Today he shares key lessons from bringing Copilot to production—including an unexpected discovery about AI hallucinations.**
+
+While reviewing user feedback for Pulumi Copilot, Pulumi's AI assistant for cloud infrastructure, one message caught my eye: "Your tool doesn't know anything!". Having just made some changes, I braced for the worst. But the evals were still looking strong, so what was going on?
```

Comment:
Here is a slightly more neutral proposed wording for the opening, LMK if you're OK with this:

"Today we want to share some insights from the Copilot team at Pulumi, where we're building an AI assistant for cloud infrastructure - including some learnings about AI hallucinations that taught us not only about AI but also about own product and our users.

While reviewing user feedback for Pulumi Copilot, our AI assistant for cloud infrastructure, one message caught our attention: "Your tool doesn't know anything!". Having just made some changes, we braced for the worst. But the evals were still looking strong, so what was going on?

The user was attempting to force-delete a stack that still contained resources. Copilot confidently suggested utilizing a `--force` flag, which would have been a logical solution... except this flag doesn't exist in Pulumi! This was an instance of pure hallucination. Such occurrences are typical when building with LLMs. However, this particular error was more than a simple glitch - it tells us something about the product. With the recent launch of our REST API and Copilot's continued operation in production, we would like to share several insights we've gained while building with LLMs."